### ISSUE TITLE
cloud foundary version 2.13 has issues while fetching appliationdetails

### ISSUE THREAD

**Issue by skmedishetty** [cloudfoundry/cf-java-client#1035]
Team,

I see our application running for 2 or 3 days we see issues **IOExcetion: connection closed premature** don't know whether  the issue with cloudfoundry reactor or cloudfoudry operations or reactor core.
Tried using latest jars with that we see issues related to memory leak

Error

**Comment by twoseat** [comment]
We're about to release a new version that has a fix that _might_ address this issue. Once it's available (should be tomorrow) it would be great if you could try it and let us know if it helps.

**Comment by skmedishetty** [comment]
@twoseat  is it a know issue ?
I see the problem is while getting oauth token or getting application details.. if the link is broken it never recovers.

**Comment by twoseat** [comment]
There are a few things that it might be, but your description is vague enough that I can't say what it is for sure. But we've done some work on memory leaks, and on token handling, that may help. And if it doesn't then at least we'll know what _isn't_ the answer!

**Comment by skmedishetty** [comment]
Let explain in details whats happening, we are trying to read the application details like number of instances using cloudfoundary API.  it works good for could of hours (sometime days).

If we have any issue related to connection it fails and won't recover.

2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_212]
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [na:1.8.0_212]
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [na:1.8.0_212]
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] at java.lang.Thread.run(Thread.java:748) [na:1.8.0_212]
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] Caused by: reactor.core.Exceptions$ReactiveException: **io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused:** api.apps.xxx.gaia.jpmchase.net/169.125.192.89:443
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] at reactor.core.Exceptions.propagate(Exceptions.java:219) ~[reactor-core-3.0.7.RELEASE.jar!/:3.0.7.RELEASE]
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:119) ~[reactor-core-3.0.7.RELEASE.jar!/:3.0.7.RELEASE]
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.Mono.block(Mono.java:1323) ~[reactor-core-3.0.7.RELEASE.jar!/:3.0.7.RELEASE]
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] at net.jpmchase.gaia.autoscaler.broker.service.ApplicationService.getApplicationInstances(ApplicationService.java:185) ~[classes/:na]
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] at net.jpmchase.gaia.autoscaler.broker.service.ApplicationService.decideAndScaleByInstance(ApplicationService.java:110) ~[classes/:na]
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] at net.jpmchase.gaia.autoscaler.broker.service.ApplicationService.consumeIdentifier(ApplicationService.java:64) ~[classes/:na]
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] ... 9 common frames omitted
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] Suppressed: java.lang.Exception: #block terminated with an error
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.BlockingSingleSubscriber.blockingGet(BlockingSingleSubscriber.java:120) ~[reactor-core-3.0.7.RELEASE.jar!/:3.0.7.RELEASE]
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] ... 13 common frames omitted
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] Caused by: io.netty.channel.AbstractChannel$AnnotatedConnectException: Connection refused: api.apps.dev.na-3a.gaia.jpmchase.net/169.125.192.89:443
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException:
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] Assembly trace from producer [reactor.core.publisher.MonoPeek] :
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] reactor.core.publisher.Mono.checkpoint(Mono.java:1400)
2020-02-18T13:05:46.953-05:00 [APP/PROC/WEB/0] [OUT] org.cloudfoundry.reactor.client.v2.spaces.ReactorSpaces.listApplications(ReactorSpaces.java:174)

**Comment by skmedishetty** [comment]
once above situation happens.. we see below errors continuously.

[OUT] Suppressed: reactor.core.publisher.FluxOnAssembly$AssemblySnapshotException: null
2020-02-18T13:13:16.952-05:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoOnAssembly.<init>(MonoOnAssembly.java:54) ~[reactor-core-3.0.7.RELEASE.jar!/:3.0.7.RELEASE]
2020-02-18T13:13:16.952-05:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.Mono.checkpoint(Mono.java:1400) ~[reactor-core-3.0.7.RELEASE.jar!/:3.0.7.RELEASE]
2020-02-18T13:13:16.952-05:00 [APP/PROC/WEB/0] [OUT] at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:268) ~[cloudfoundry-client-reactor-2.16.0.RELEASE.jar!/:2.16.0.RELEASE]
2020-02-18T13:13:16.952-05:00 [APP/PROC/WEB/0] [OUT] at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.invalidate(AbstractUaaTokenProvider.java:112) ~[cloudfoundry-client-reactor-2.16.0.RELEASE.jar!/:2.16.0.RELEASE]
2020-02-18T13:13:16.952-05:00 [APP/PROC/WEB/0] [OUT] at org.cloudfoundry.reactor.util.AbstractReactorOperations.lambda$invalidateToken$23(AbstractReactorOperations.java:277) ~[cloudfoundry-client-reactor-2.16.0.RELEASE.jar!/:2.16.0.RELEASE]
2020-02-18T13:13:16.952-05:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoThenMap$ThenMapMain.onNext(MonoThenMap.java:120) ~[reactor-core-3.0.7.RELEASE.jar!/:3.0.7.RELEASE]
2020-02-18T13:13:16.952-05:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onNext(FluxOnErrorResume.java:72) ~[reactor-core-3.0.7.RELEASE.jar!/:3.0.7.RELEASE]
2020-02-18T13:13:16.952-05:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:962) ~[reactor-core-3.0.7.RELEASE.jar!/:3.0.7.RELEASE]
2020-02-18T13:13:16.952-05:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.MonoThenMap$ThenMapMain$ThenMapInner.onNext(MonoThenMap.java:237) ~[reactor-core-3.0.7.RELEASE.jar!/:3.0.7.RELEASE]
2020-02-18T13:13:16.952-05:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxDoFinally$DoFinallySubscriber.onNext(FluxDoFinally.java:124) ~[reactor-core-3.0.7.RELEASE.jar!/:3.0.7.RELEASE]
2020-02-18T13:13:16.952-05:00 [APP/PROC/WEB/0] [OUT] at reactor.core.publisher.FluxPeek$PeekSubscriber.onNext(FluxPeek.java:173) ~[reactor-core-3.0.7.RELEASE.jar!/:3.0.7.RELEASE]

**Comment by skmedishetty** [comment]
@twoseat  this is really important for our project, this is used to scale up/down for one our critical application. it would be really great if you can help resolving this issue

**Comment by skmedishetty** [comment]
context provider channel pools size increases as time goes on (this is is with cloufoundary **4.2.x**
**size = 117**
{PooledConnectionProvider$PoolKey@16805}  -> {SimpleFifoPool@16806} {PooledConnectionProvider$PoolKey@16807}  -> {SimpleFifoPool@16808} 
Might be this is the main cause for memory leak.
I'm in a situation where unable to use old versions 2.x or 4.2.x with two different issues.

**Comment by twoseat** [comment]
Can you try with version `4.4.0` of the client and let me know if you see the same problems?

**Comment by skmedishetty** [comment]
@twoseat  I used 4.2.0 Release version which is the latest available in our repository.
It also has same issue.

**Comment by violetagg** [comment]
@skmedishetty Can you enable `reactor.netty` logs?

Can you tell me how many times you see the log below:

```
Creating a new fixed client pool with name ...
```

and also can you share the log like this:

```
20:21:10.686 [reactor-http-nio-2] DEBUG r.netty.channel.BootstrapHandlers - [id: 0xa44fee86] Initialized pipeline DefaultChannelPipeline{(reactor.left.loggingHandler = io.netty.handler.logging.LoggingHandler), (BootstrapHandlers$BootstrapInitializerHandler#0 = reactor.netty.channel.BootstrapHandlers$BootstrapInitializerHandler), (reactor.left.httpCodec = io.netty.handler.codec.http.HttpClientCodec), (reactor.right.reactiveBridge = reactor.netty.channel.ChannelOperationsHandler)}
```

**Comment by twoseat** [comment]
You need to request an update to your repository as we don't provide assistance for older versions. We take great care to retain compatibility between versions, so the upgrade should be trivial.

**Comment by skmedishetty** [comment]
@twoseat  sure I will upgrade to 4.4.0 and test.

**Comment by skmedishetty** [comment]
@violetagg 

I see 3 new pool Connections are created every request.

2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] Caused by: java.lang.IllegalArgumentException: promise already done: DefaultChannelPromise@2a73f8de(failure: io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 1580 byte(s) of direct memory (used: 10484565, max: 10485760))
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.AbstractChannelHandlerContext.isNotValidPromise(AbstractChannelHandlerContext.java:883) ~[netty-transport-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] Suppressed: reactor.core.publisher.FluxOnAssembly$OnAssemblyException:
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] Assembly trace from producer [reactor.core.publisher.MonoProcessor] :
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] reactor.core.publisher.Mono.checkpoint(Mono.java:1854)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] Error has been observed at the following site(s):
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.client.v2.spaces.ReactorSpaces.listApplications(ReactorSpaces.java:174)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_ Mono.checkpoint ⇢ at org.cloudfoundry.operations.applications.DefaultApplications.get(DefaultApplications.java:259)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_____ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_____ Mono.checkpoint ⇢ at org.cloudfoundry.operations.applications.DefaultApplications.get(DefaultApplications.java:259)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_________ Mono.checkpoint ⇢ at org.cloudfoundry.operations.applications.DefaultApplications.get(DefaultApplications.java:259)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_____________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_____________ Mono.checkpoint ⇢ at org.cloudfoundry.operations.applications.DefaultApplications.get(DefaultApplications.java:259)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_____________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_____________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_____________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_________________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_____________________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_________________________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_____________________________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_________________________________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_____________________________________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_________________________________________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_____________________________________________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_________________________________________________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_____________________________________________________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:41:06.722-05:00 [APP/PROC/WEB/0] [OUT] |_________________________________________________________________________________ Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)

**Comment by skmedishetty** [comment]
Mono.checkpoint ⇢ at org.cloudfoundry.reactor.tokenprovider.AbstractUaaTokenProvider.token(AbstractUaaTokenProvider.java:255)
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:377) ~[netty-transport-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363) ~[netty-transport-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:355) ~[netty-transport-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:102) ~[netty-codec-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) ~[netty-transport-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:316) ~[netty-codec-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:303) ~[netty-codec-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:417) ~[netty-codec-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:271) ~[netty-codec-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251) ~[netty-transport-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:227) ~[netty-handler-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1470) ~[netty-handler-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1219) ~[netty-handler-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1266) ~[netty-handler-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:493) ~[netty-codec-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:432) ~[netty-codec-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[netty-transport-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[netty-transport-4.1.44.Final.jar!/:4.1.44.Final]
2020-02-19T14:43:06.798-05:00 [APP/PROC/WEB/0] [OUT] at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:792) ~[netty-transport-native-epoll-4.1.44.Final-linux-

**Comment by skmedishetty** [comment]
@ I don't see logs related to handler while error, but when it works properly i see below.
2020-02-19 18:09:40.815 DEBUG 16 --- [client-epoll-11] reactor.netty.channel.BootstrapHandlers  : [id: 0xb9e623c3] Initialized pipeline DefaultChannelPipeline{(reactor.left.sslHandler = io.netty.handler.ssl.SslHandler), (reactor.left.loggingHandler = io.netty.handler.logging.LoggingHandler), (reactor.left.sslReader = reactor.netty.tcp.SslProvider$SslReadHandler), (BootstrapHandlers$BootstrapInitializerHandler#0 = reactor.netty.channel.BootstrapHandlers$BootstrapInitializerHandler), (reactor.left.httpCodec = io.netty.handler.codec.http.HttpClientCodec), (reactor.left.decompressor = io.netty.handler.codec.http.HttpContentDecompressor), (reactor.right.reactiveBridge = reactor.netty.channel.ChannelOperationsHandler)}

**Comment by violetagg** [comment]
@skmedishetty how do you enable logs?
@twoseat do you use somewhere `wiretap(...)`

**Comment by twoseat** [comment]
@violetagg https://github.com/cloudfoundry/cf-java-client/blob/master/cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/_DefaultConnectionContext.java#L278 is the only use.

**Comment by skmedishetty** [comment]
@violetagg  I have enabled debug root to see the logs of all package.

**Comment by violetagg** [comment]
> @violetagg https://github.com/cloudfoundry/cf-java-client/blob/master/cloudfoundry-client-reactor/src/main/java/org/cloudfoundry/reactor/_DefaultConnectionContext.java#L278 is the only use.

Do you have this enabled always?
This might be related to this issue (currently in 0.9.5.BUILD-SNAPSHOT)
https://github.com/reactor/reactor-netty/commit/643c6725486ee8e95b7e805ab8d4178feacf2b8c

**Comment by twoseat** [comment]
@violetagg It's always set that way, but whether it's enabled or not depends on what the user has logging set to doesn't it?

**Comment by violetagg** [comment]
@skmedishetty ok then this fix might be related

**Comment by skmedishetty** [comment]
@violetagg  ok

@violetagg / @twoseat 

While I was troubleshooting logs i found this today.

1.  While getting application details  saw timeouts continuously **cloudFoundryOperations.applications().get(GetApplicationRequest.builder().name(applicationName).build())
                .timeout(Duration.of(AutoScalerConstants.TIME_MILLISECONDS, ChronoUnit.MILLIS))
                .retry(3)
                .doOnSuccess(s -> LOGGER.info("Successfully Getting Application Details {}", applicationName+s.getInstances()))
                .doOnError(e -> LOGGER.info("UnSuccessful Getting Application Details {}", applicationName))
                .block(Duration.of(AutoScalerConstants.TIME_MILLISECONDS, ChronoUnit.MILLIS));**
2 then after an hour I saw below error 
[client-epoll-11] r.n.http.client.HttpClientOperations     : [] Received response (auto-read:false) : [Cache-Control=no-cache, no-store, max-age=0, must-revalidate, Cache-Control=no-store, Content-Type=application/json;charset=UTF-8, Date=Wed, 19 Feb 2020 18:23:26 GMT, Expires=0, Pragma=no-cache, Pragma=no-cache, Strict-Transport-Security=max-age=31536000 ; includeSubDomains, Www-Authenticate=Bearer error="invalid_token", error_description="Invalid refresh token expired at Wed Feb 19 18:00:42 UTC 2020", X-Content-Type-Options=nosniff, X-Frame-Options=DENY, X-Vcap-Request-Id=b9fb48f5-87a5-448d-4608-7f6f37ebdb25, X-Xss-Protection=1; mode=block, Content-Length=109]
3. After all this when I put debug point 

**promise already done: DefaultChannelPromise@2a73f8de(failure: io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 1580 byte(s) of direct memory (used: 10484565, max: 10485760))**


So my Question is in case of error is it possible to revalidate the token and refresh context ?

**Comment by skmedishetty** [comment]
@twoseat  same issue with 4.4 looks more related to netty

**Comment by skmedishetty** [comment]
@violetagg  @twoseat  I upgraded 0.9.5 reactive-netty and cloud-foundry to 4.4.0 now I don't see memory issue but see below error 
org.cloudfoundry.reactor.util.Operator$ResponseReceiver$InvalidTokenException
It runs for couple of hours and I see this issue.

Do you guys have idea why it fails after couple of hours ?

**Comment by dmikusa-pivotal** [comment]
For what it's worth, there is probably a token refresh problem happening. I believe we've fixed at least one bug in this area since 4.4.0. If you haven't upgraded, it would be worth giving that a try.

I'm going to close this since the issue is very old now. If you're still having issues even on the latest version of the client, please open a new issue and we can examine further.
